graphs/adapter-jgrapht/src/main/kotlin/com/anaplan/engineering/azuki/graphs/adapter/jgrapht/declaration/GraphDeclarationBuilderFactory.kt

```diff
@@ -1,43 +1,59 @@
+class UndirectedGraphDeclarationBuilderFactory : JGraphTDeclarationBuilderFactory<UndirectedGraphDeclaration<*>> {
 
-    override val declarationClass = GraphDeclaration::class.java
+    override val declarationClass = UndirectedGraphDeclaration::class.java
 
-    override fun create(declaration: GraphDeclaration<*>): JGraphTDeclarationBuilder<GraphDeclaration<*>> =
+    override fun create(declaration: UndirectedGraphDeclaration<*>): JGraphTDeclarationBuilder<UndirectedGraphDeclaration<*>> =
         GraphDeclarationBuilder(declaration)
 
-    private class GraphDeclarationBuilder(declaration: GraphDeclaration<*>) :
-        JGraphTDeclarationBuilder<GraphDeclaration<*>>(declaration) {
+}
 
-        override fun build(env: ExecutionEnvironment) {
-            val graph = if (declaration.directed) {
-                SimpleDirectedGraph<Any, DefaultEdge>(DefaultEdge::class.java)
-            } else {
-                SimpleGraph<Any, DefaultEdge>(DefaultEdge::class.java)
-            }
-            declaration.vertices.map {
-                graph.addVertex(it)
-            }
-            declaration.edges.map {
-                if (it.first == it.second) {
-                    throw LateDetectUnsupportedActionException("JgraphT cannot have a loop back to the same vertex by default")
-                }
-                // necessary to add vertices before using in an edge
-                graph.addVertex(it.first)
-                graph.addVertex(it.second)
-
-                graph.addEdge(it.first, it.second)
-            }
-            env.addGraph(declaration.name, graph)
+class DirectedGraphDeclarationBuilderFactory : JGraphTDeclarationBuilderFactory<DirectedGraphDeclaration<*>> {
+
+    override val declarationClass = DirectedGraphDeclaration::class.java
+
+    override fun create(declaration: DirectedGraphDeclaration<*>): JGraphTDeclarationBuilder<DirectedGraphDeclaration<*>> =
+        GraphDeclarationBuilder(declaration)
+
+}
+
+private class GraphDeclarationBuilder<T : GraphDeclaration<*>>(declaration: T) :
+    JGraphTDeclarationBuilder<T>(declaration) {
+
+    override fun build(env: ExecutionEnvironment) {
+        var graph: Graph<Any, DefaultEdge> = when (declaration) {
+            is UndirectedGraphDeclaration<*> -> GraphTypeBuilder.undirected<Any, DefaultEdge>()
+                .edgeClass(DefaultEdge::class.java).buildGraph()
+
+            is DirectedGraphDeclaration<*> -> GraphTypeBuilder.directed<Any, DefaultEdge>()
+                .edgeClass(DefaultEdge::class.java).buildGraph()
+
+            else -> throw LateDetectUnsupportedActionException("Unkown graph declaration type ${declaration::class}")
         }
+        declaration.vertices.map {
+            graph.addVertex(it)
+        }
+        declaration.edges.map {
+            if (it.first == it.second) {
+                graph = GraphTypeBuilder.forGraph(graph).allowingSelfLoops(true).buildGraph()
+            }
```

Comment:
⛏️ To avoid rebuilding the graph over and over again, I would prefer
```
        var graphBuilder = when (declaration) {
            is UndirectedGraphDeclaration<*> -> GraphTypeBuilder.undirected<Any, DefaultEdge>()
                .edgeClass(DefaultEdge::class.java)
            is DirectedGraphDeclaration<*> -> GraphTypeBuilder.directed<Any, DefaultEdge>()
                .edgeClass(DefaultEdge::class.java)
            else -> throw LateDetectUnsupportedActionException("Unkown graph declaration type ${declaration::class}")
        }
        if (declaration.edges.any { it.first == it.second }) graphBuilder.allowingSelfLoops(true)
        val graph = graphBuilder.buildGraph()
```